src/Microsoft.OData.Core/Json/PooledByteBufferWriter.cs

```diff
@@ -0,0 +1,221 @@
+        public const int MaximumBufferSize = 0X7FFFFFC7;
+
+        private PooledByteBufferWriter()
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="PooledByteBufferWriter"/> to which data can be written,
+        /// with a specified initial capacity.
+        /// </summary>
+        /// <param name="initialCapacity"></param>
+        public PooledByteBufferWriter(int initialCapacity) : this()
+        {
+            Debug.Assert(initialCapacity > 0);
+
+            this.rentedBuffer = ArrayPool<byte>.Shared.Rent(initialCapacity);
+            this.index = 0;
+        }
+
+        /// <summary>
+        /// Gets a <see cref="ReadOnlyMemory{byte}"/> that contains the data
+        /// written to the underlying buffer so far.
+        /// </summary>
+        public ReadOnlyMemory<byte> WrittenMemory
+        {
+            get
+            {
+                Debug.Assert(this.rentedBuffer != null);
+                Debug.Assert(this.index <= this.rentedBuffer.Length);
+                return this.rentedBuffer.AsMemory(0, this.index);
+            }
+        }
+
+        /// <summary>
+        /// Gets the amount of data written to the underlying buffer.
+        /// </summary>
+        public int WrittenCount
+        {
+            get
+            {
+                Debug.Assert(rentedBuffer != null);
+                return this.index;
+            }
+        }
+
+        /// <summary>
+        /// Gets the total amount of space within the underlying buffer.
+        /// </summary>
+        public int Capacity
+        {
+            get
+            {
+                Debug.Assert(this.rentedBuffer != null);
+                return this.rentedBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        /// Gets the amount of available space that can be written to
+        /// without forcing the underlying buffer to grow.
+        /// </summary>
+        public int FreeCapacity
+        {
+            get
+            {
+                Debug.Assert(this.rentedBuffer != null);
+                return this.rentedBuffer.Length - this.index;
+            }
+        }
+
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        /// <param name="count"></param>
+        public void Advance(int count)
+        {
+            Debug.Assert(this.rentedBuffer != null);
+            Debug.Assert(count >= 0);
+            Debug.Assert(this.index <= this.rentedBuffer.Length - count);
+            this.index += count;
+        }
+
+        /// <summary>
+        /// <inheritdoc/>
+        /// </summary>
+        /// <param name="sizeHint"></param>
+        /// <returns></returns>
+        public Memory<byte> GetMemory(int sizeHint = MinimumBufferSize)
+        {
+            CheckAndResizeBuffer(sizeHint);
+            return this.rentedBuffer.AsMemory(index);

```

Comment:
I assume this method returns the segment of the array that hasn't been written to...

Reply:
Yes. The next position to write to is tracked by the `index`.